library/ssl_tls.c

```diff
@@ -58,6 +58,67 @@
 
 #define PROPER_HS_FRAGMENT 0x75555555
 
+#if defined(MBEDTLS_VALIDATE_SSL_KEYS_INTEGRITY)
+static uint32_t mbedtls_hash( const void *data, size_t data_len_bytes )
+{
+    uint32_t result = 0;
+    size_t i;
+    /* data_len_bytes - only multiples of 4 are considered, rest is truncated */
+    for( i = 0; i < data_len_bytes >> 2; i++ )
+    {
+        result ^= ( (uint32_t*) data )[i];
+    }
+    return result;
+}
+#endif
+#if defined(MBEDTLS_SSL_TRANSFORM_OPTIMIZE_CIPHERS)
+static int mbedtls_ssl_switch_key( mbedtls_ssl_transform *transform,
+                                   const mbedtls_operation_t operation )
+{
+    unsigned char * key;
+    int ret;
+#if defined(MBEDTLS_VALIDATE_SSL_KEYS_INTEGRITY)
+    uint32_t hash;
+#endif
+    if( operation == MBEDTLS_ENCRYPT )
+    {
+        key = transform->key_enc;
+#if defined(MBEDTLS_VALIDATE_SSL_KEYS_INTEGRITY)
+        hash = transform->key_enc_hash;
+#endif
+    }
+    else if ( operation == MBEDTLS_DECRYPT )
+    {
+        key = transform->key_dec;
+#if defined(MBEDTLS_VALIDATE_SSL_KEYS_INTEGRITY)
+        hash = transform->key_dec_hash;
+#endif
+    }
+    else
+    {
+        return ( MBEDTLS_ERR_SSL_INTERNAL_ERROR );
+    }
+#if defined(MBEDTLS_VALIDATE_SSL_KEYS_INTEGRITY)
+    /* Check hash */
+    if( hash != mbedtls_hash( key, transform->key_bitlen >> 3 ) )
```

Comment:
@AndrzejKurek , it seems to be problematic. We are checking keys integrity only in case we are going to switch enc/dec mode. In other cases round keys are not checked and they were generated before this check. 
As we need integrity check only for ccm mode I propose to check check it as a part of AES if cmm mode does not context switching at all.

Reply:
We are checking the hash each time the `switch_key` function is called, and it is called before each encryption/decryption. There are no other cases that make the key being used without a hash check.

Reply:
@AndrzejKurek , we are checking raw keys, but if we  are not performing switch (the operation was not changed from encrypt to decrypt for example ), the aes context will use previously created  round keys and not raw key that saved in ssl context

Reply:
So we should check raw keys if we are performing switching and round keys if we are not